### PR TITLE
make model closer to scalameta ast & create model for generics 

### LINKED ISSUE #3: Create Model for generics

### LINKED ISSUE #6: make model as close to meta as possible
I would like to see this project help the efforts in scala meta and meta doc if in any way possible. One is to keep the trees as similar as possible, so that it is easy to take over code between the projects or interop as libraries.

It seems to me that the strategy should be that we replicate meta's trees in their current state, but add all doc relevant data that meta computes using a compiler as actual data into the model, e.g. super classes.

It sounds like metadoc is doing that (without having checked that). If that is the case, we can basically take a subset of metadoc's model as we go and eventually hopefully the full model and concentrate on the backends and improving metadocs model rather than newly inventing it from scratch.

So we should adapt the current model to metadoc's model (if it is true that it is designed as described above).

### PR DESCRIPTION
**Pull request by kolowheel**
fixes #3  
fixes #6

### CODE REVIEW

**Comment by kolowheel** [comment]
New class Name is an unique identifier for  DocElement.
Elements those are could be documented should implement trait DocElement.
Rest model members should extend trait Tree.
Unresolved questions:
How should default value be implemented? (there is default: Option[Term] filed in original scalameta tree, but in doc we don't need all staff connected with Term)
Should we have file:SourceFile member in Defn? 
Should we have parent reference in tree trait?{trait tree {def parent : Tree }}

**Comment by cvogt** [comment]
@kolowheel 
For all the things you are not sure about if we need them? Leave them out for now, we can add them when we need them.
Not sure what you mean by map Type.Name => Name => DocElement 

Can you explain more?

**Comment by cvogt** [comment]
do we still need DocElement? Can't we use something from the meta trees instead? E.g. Tree or Symbol or something?

File: src/main/scala/DocGenerator.scala [hunk 1]
```diff
@@ -96,19 +98,19 @@ object LatexDocGenerator extends DocGenerator {
...
     s"""\\subsection{Method summary}{
       \\begin{verse}
-        ${mehtods.collect { case e: MethodDoc => s"{\\bf def ${e.name}(${dumpMethodInputs(e)})}\\\\" }.mkString("\n")}
+        ${mehtods.collect { case e: Def => s"{\\bf def ${e.name}(${dumpMethodInputs(e)})}\\\\" }.mkString("\n")}
       \\end{verse}
       }""" 
   }
 
-  def dumpMethodInputs(e: MethodDoc): String = e.inputs.map(_.paramType.name).mkString(",")
+  def dumpMethodInputs(e: Def): String = e.inputs.map(_.decltpe.asInstanceOf[Type.Name]).mkString(",")
```
Comment (clhodapp) [inline src/main/scala/DocGenerator.scala]: This cast looks a bit sketchy... Is it a `Type` or a `Name`?

File: src/main/scala/DocGenerator.scala [hunk 2]
```diff
@@ -96,19 +98,19 @@ object LatexDocGenerator extends DocGenerator {
       }}""" 
   }
 
-  def processMethodsSummary(mehtods: Seq[DocElement]): String = {
+  def processMethodsSummary(mehtods: Seq[Tree]): String = {
```
Comment (cvogt) [inline src/main/scala/DocGenerator.scala]: typo mehtods :)

**Comment by cvogt** [comment]
@kolowheel  heading the right way. Can we still move ClassDoc, TraitDoc into the meta structure? Probably means adding Template. Or anything that prevents it?

**Pull request merged**